### PR TITLE
Ruby: expand DataFlow API

### PR DESCRIPTION
**Pull request by asgerf**
Adds some classes to do basic stuff with `DataFlow` nodes without having to map down to CFG and AST nodes, and a few other tidbits I found myself missing.

To ensure some degree of battle-hardening I (mostly) ported three library models to use the new nodes where it makes sense. The affected models are `Rails.qll`, `Railties.qll`, and `ActionController.qll`.

Migrating from AST nodes to DataFlow nodes revealed a couple of pre-existing issues, some of which have now been fixed or fixed enough to avoid a regression:
- https://github.com/github/codeql/pull/10928
- https://github.com/github/codeql/pull/10918

Some other issues which were too hard to fix prior to opening this PR:
- We're missing flows into captured variables in some cases, in particular when returning a lambda. I attempted a fix in [this draft PR](https://github.com/github/codeql/pull/11010) which would align closer with how we do things in JS, but after a whole day of debugging strange call graph changes, I've put that aside for now. Instead, I've resorted to a rather unambitious workaround in https://github.com/github/codeql/pull/11024/commits/ab91b5772bcfb72084dd83523de1abac43f24942.
- Blocks are assumed to capture `self`, but of course that's not always the case. There is currently no way to tell the analysis what the value of `self` is, so we have to work around it for now. This means using things like `getParent+()` and treating the module `self` as if it's an instance of the module.

### getConst() and what about API graphs?

One of the changes that may look odd is the introduction of `DataFlow::getConst()` and `ConstRef`. It provides convenient access to constants that may resolve to a given qualified name. For example, the incantation
```codeql
DataFlow::getConst("ActionController").getConst("Base").getADescendentModule()
```
gives us all descendents of `ActionController::Base`.

This was probably the biggest use case for API graphs in Ruby, as you can get to (almost) the same thing with `API::getTopLevelMember("ActionController").getMember("Base")`. So why add `getConst()`?

- `getConst()` doesn't rely on the call graph, which means it can be used for things like `getACallSimple()` and other pre-call graph customizations (we have quite a few of those in JS and I expect we'll find the need in Ruby as well).
- I intend to share more of the API graphs implementation between languages (stay tuned!), but the rules for constant lookup in Ruby are pretty language-specific, so it would be nice to have some of this factored out anyway.
- `getConst()` handles the rules of constant lookup more precisely than API graphs, in particular when related to `include`, nested modules, and modules with more than one declaration. I think API graphs could be extended to make this work, but it's not really a natural fit for its use/def tracking, it would need to be a bunch of new predicates specific to constant lookup, which might as well exist earlier in the pipeline.

I replaced some uses of API graphs with `getConst`, not because it matters in practice, I just had to ensure `getConst()` is being used enough to test it in the wild.

I hope the name `getConst()` makes more sense in the context of Ruby. I think the name in API graphs, `getMember`, was used mostly because that's what JS called it.

### Evaluation

[Evaluation](https://github.com/github/codeql-dca-main/issues/8214) shows
- 100 new taint sinks
- 491 new taint sources, resulting in 1890 new tainted expressions
- 3 new call edges due to the change in `flowsTo`
- There seems to be a slight performance hit, which I might investigate a bit further, but I wanted to get the PR up for now.

The new sources and sinks are mainly due to resolving methods upwards in the hierarchy in `ActionController`. The combination `getADescendentModule().getAnInstanceMethod()` takes "sideways" method resolution into account without having to really think about it, so I like that.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/ast/Constant.qll
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Constant.qll:175]: It feels a bit strange that the "inverse" of `getInt` is named `getInt` . Wouldn't `ConstantValue::fromInt` be a better name?

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Module.qll:27]: This change is very similar to https://github.com/github/codeql/pull/10747/files#diff-9333de60c5c1b8283b484ae3b7e3cbef24b7bf52aee695df420d0e3f4d55403b 

The main difference is that you use `*` instead of `+` which would make all modules ancestors of themselves, which is odd (even though it is handy in some cases).

@hvitved suggested a `pragma[inline]` annotation on `getAnAncestor` , perhaps you should do that as well.

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/ast/Module.qll:157]: I'm not sure I understand what a "canonical" enclosing module means. Could you give an example of a non-canonical enclosing module?

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:285]: You can have assignments like
```
obj[a, b, c] = x
```
I'm not sure we want to capture such things in this predicate though. Just wanted to make sure you ignored such calls on purpose.

**Review by aibaars — COMMENTED** [review]
Partial review.

File: ruby/ql/lib/codeql/ruby/ast/Constant.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Constant.qll:175]: `fromInt` sounds good to me 👍

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll:27]: > which would make all modules ancestors of themselves, which is odd (even though it is handy in some cases)

I think it's quite common to define things like this, for example, a type is usually considered a subtype of itself. Also a module is technically on its own ancestor chain.
```rb
module A
end
A.ancestors # [A]
```

I also tend to prefer `*` because you can easily get the `+` variant by doing `getADirectAncestor().getAnAncestor()`. 

However, if the predicate only exposes the `+` variant, you have to do something like `[x, x.getAnAncestor()]` to get the `*` variant, and even that doesn't work if `x` has type `Class`. Either way it can't appear in the middle of an expression chain, which is important iMO.

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll:157]: A module can aliased, i.e. reachable via more than one chain of constants. One of these chains of constants is considered to be the "canonical" name, the one used in the `TResolved` constructor, and the rest are just aliases. This predicate gets the enclosing module _as it appears in the canonical name_.

For the module `A::X` below, the module `A` would be its canonical enclosing module.
```rb
module A
end
module B
  module A::X
  end
end
```

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:285]: It's not intended to match anything that wouldn't work for `Array` or `Hash`.

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll:27]: You're right. I somehow thought that `A.ancestors` would not be reflexive.

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll:157]: The problem is that the "canonical" name in `TResolved`  is rather arbitrary and does not always reflect what the Ruby interpreter would report as the "name" . So it may be a bit misleading. In the interpreter a module's name appears to be determined by the name of the first constant to which it is assigned.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:719]: Can you explain your thinking behind making this an AST node and not a `DataFlow::Node`? I note that JS's `DataFlow::ClassNode` extends `DataFlow::Node`.

**Review by nickrolfe — COMMENTED** [review]
Not a thorough review, but a few things I noticed. In general, I think this is a great change that should let us simplify a lot of awkward code.

Before this is merged, please make sure to add a release note describing not only the new classes, but the deprecation of `ActionControllerControllerClass`.

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll:157]: Is there another name we could use for this concept that you'd find more intuitive?

Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll:157]: That is a very good question. I couldn't find a name that is used in the Ruby documentation or implementation for the "parent" or "container" of a module/class.

There does not seem to be any reflection method to get that information in the standard library (apart from parsing the value of `Module.name`). However, there is a method named `module_parent` in Rails: https://apidock.com/rails/v6.0.0/Module/module_parent

So perhaps that is the best name for this concept?

File: ruby/ql/lib/codeql/ruby/ast/Constant.qll
Comment (aibaars) [reply ruby/ql/lib/codeql/ruby/ast/Constant.qll:175]: Same of the other getters, of course.

Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Constant.qll]: Renamed to `fromInt` etc

File: ruby/ql/lib/codeql/ruby/ast/Module.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ast/Module.qll]: Pushed a commit to rename:
- `getCanonicalEnclosingModule` -> `getParentModule`
- `getCanonicalNestedModule` -> `getNestedModule`

**Comment by asgerf** [comment]
I've pushed 0f1b348, which [fixed the regression in metasploit](https://github.com/github/codeql-dca-main/issues/8229).

**Comment by nickrolfe** [comment]
I've been doing some Rails modelling based on this branch, and I'm wondering about this scenario:

```rb
class Base
end

class Foo < Base
end

class Bar < Foo
end
```

`DataFlow::getConst("Base").getADescendentModule()` gives me `Base`, `Foo`, and `Bar`, but I couldn't see an obvious way of excluding `Base` to get only `Foo` and `Bar`.

(I'm trying to find methods on classes derived from a Rails base class, but I don't want to find methods on that base class itself when we're analysing `rails/rails`).

**Comment by asgerf** [comment]
@nickrolfe on that snippet it only gives me `Foo` and `Bar`. I'm guessing there must be something else going on in `rails/rails` that's causing this behaviour?

**Comment by nickrolfe** [comment]
Ah. Thanks, I'll do some digging into Rails.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:949]: `this.`

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1069]: `CallNode` instead of `ExprNode`?

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1104]: cache?

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1223]: ql doc is out of sync

**Review by hvitved — CHANGES_REQUESTED** [review]
Initial review; overall LGTM.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:123]: The underlying bug is that SSA doesn't consider a lambda expression to escape when returned, as highlighted by the added test case. As mentioned in the PR description, the code here is a workaround to avoid regression in the short-term, and I'm not really interested in generalizing the workaround. I initially tried landing such fixes prior to opening this PR, but after a while I didn't want to block the PR any longer.

Obviously the bug in SSA should be fixed, but at the architectural level I don't think this layer ought care about it. I'd like to post-pone flow-sensitive handling of captured variables until after the call graph, so the layers would look something like:
- Materialize data-flow nodes.
- Compute flow-insensitive step relation, treating captured variables flow-insensitively (basically SSA only used for true locals)
- Compute call graph, using the flow-insensitive relation for propagating type information
- Compute flow-sensitive step relation (for use by DataFlow::Configuration etc). This stage can then in principle benefit from seeing the call graph.

Flow sensitivity doesn't matter very much for propagating type information, whereas it matters a lot more to taint tracking and the like. Using a single handling of captured variables to cover both isn't optimal.

I tried to make a start on this [here](https://github.com/github/codeql/pull/11010) but as mentioned, I had to shelf it to make some progress on this PR.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:554]: I haven't been able to figure out how get the last CFG node in a `Namespace`, but if you can write up the relevant changes to the SSA library I'm happy to include it in the PR and then remove the new data-flow node.

File: ruby/ql/test/library-tests/dataflow/global/captured_variables.rb
Comment (asgerf) [reply ruby/ql/test/library-tests/dataflow/global/captured_variables.rb:1]: The test includes calls to unknown methods, which makes it a bit more tricky to make the test runnable.

File: ruby/ql/test/library-tests/dataflow/global/captured_variables.rb
Comment (asgerf) [reply ruby/ql/test/library-tests/dataflow/global/captured_variables.rb:9]: AFAICT it's an issue with capture steps in SSA

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:901]: Yes, thanks for catching that

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1005]: We do this in JavaScript so I don't see why we'd need to hide it in Ruby.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1098]: Not being a `CallNode` was deliberate. It exposes member predicates that aren't necessarily helpful or do what you expect when thinking about a hash/array literal.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1069]: same

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:1285]: We already have `Module#getASubClass()` from before this PR, which returns an immediate subclass (not transitive). For that reason I'm trying to avoid adding things called `getSubClass/getSubclass` that don't behave the same way.

**Comment by asgerf** [comment]
> On the naming of `getConst`, is there any reason not to use the full word `getConstant`? 

I preferred a short name given how ubiquitous the corresponding `DataFlow::moduleImport` and `DataFlow::globalVarRef` are in JavaScript. I'm not sure it's a good enough reason though, since it's not that much shorter. I'd be happy to hear if anyone else has thoughts on this as well.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPublic.qll:302]: Factored out `getAnElementWriteCall` to reduce duplication

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:123]: Makes perfect sense; thanks.

**Comment by asgerf** [comment]
[New evaluation](https://github.com/github/codeql-dca-main/issues/8390) shows we gain about 1,600 new call edges due to handling of `proc`, and is otherwise unchanged.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll]: Merged your fix from https://github.com/asgerf/codeql/pull/1, thanks!

**Review by hvitved — APPROVED** [review]
Let's see what DCA says.

**Comment by asgerf** [comment]
[Latest evaluation](https://github.com/github/codeql-dca-main/issues/8423) still looks good

**Review by hvitved — APPROVED** [review]

**Pull request merged**
